### PR TITLE
untangle bpf_host

### PR DESCRIPTION
**Pull request by julianwiedmann**
The shared code for from-host / from-netdev paths makes it unnecessarily difficult to follow along. And actually only very little of the code is shared. Split it up to make things easier.

Probably easiest to review patch-by-patch.

There's more metrics with questionable direction in these paths. But this is a start.

```release-note
bpf: split up from-host / from-netdev paths in bpf_host
```

### CODE REVIEW

**Comment by julianwiedmann** [comment]
/test-1.26-net-next

**Comment by joestringer** [comment]
The release note for this is "release-note/minor". While this does look like an important and considerable change to the datapath, the "release-note" is intended as a user-facing note about the changes. I don't think that any users are likely to be able to interpret what "split up from-host / from-netdev paths in bpf_host" means. If you think there is something directly user-facing about this change, then please update the release note to reflect that impact. If not, then let's change the release-note to release-note/misc.

**Review by NikAleksandrov — APPROVED** [review]

File: bpf/bpf_host.c
Comment (pchaigno) [inline bpf/bpf_host.c:248]: This first commit sounds like it's fixing a bug. We should move it to its own pull request and backport it to v1.12, unless we think it can't affect users for some other reason.

File: bpf/bpf_host.c
Comment (pchaigno) [inline bpf/bpf_host.c:1075]: I think the split is a great improvement, but I'm a bit worried of the amount of backporting conflicts it will cause for v1.13. Maybe best to wait the release before doing that?

File: bpf/bpf_host.c
Comment (pchaigno) [inline bpf/bpf_host.c:1070]: This is also a bug fix, which could be in its own pull request and backported to v1.12, maybe with other fixes for incorrect directions.

**Review by pchaigno — CHANGES_REQUESTED** [review]
The changes themselves look great, but I think we'll need to split them into different PRs and we may want to wait for the release before merging such a large code move.

If some of the bugfixes depend on the refactoring, then maybe it's fine to not backport those bugfixes as they are not fixing major bugs. That doesn't look to be the case though.

File: bpf/bpf_host.c
Comment (julianwiedmann) [reply bpf/bpf_host.c:248]: It strikes me as a fairly exotic bug, but sure why not.

File: bpf/bpf_host.c
Comment (julianwiedmann) [reply bpf/bpf_host.c:1075]: You mean delay merging the split-up patches into master, until v1.13 has been released? Sure, one more time through rebase/CI hell won't make a difference ;).

File: bpf/bpf_host.c
Comment (julianwiedmann) [reply bpf/bpf_host.c:1070]: Yep, can bundle up the two direction fixes into a PR. Don't think I'll be looking at more incorrect directions before the split is in.

File: bpf/bpf_host.c
Comment (pchaigno) [reply bpf/bpf_host.c:248]: Well, what's the exact impact? I haven't checked and there was no discussion of that in the commit description :wink:

File: bpf/bpf_host.c
Comment (pchaigno) [reply bpf/bpf_host.c:1075]: That one more time for you means one less time for every single backport that touches that code, so probably a win :grin:

File: bpf/bpf_host.c
Comment (julianwiedmann) [reply bpf/bpf_host.c:248]: > Well, what's the exact impact?

Just from looking at the code? Your favorite - asymmetry ;). We can miss cases (`NO_REDIRECT && !ENABLE_HOST_ROUTING`) where a packet **needs** to go from-netdev-to-pod via the stack, as (1) iptables is active and (2) the from-pod-to-netdev path will pass replies via the stack (due to `!ENABLE_HOST_ROUTING`).

For which of those allow-listed ICMPv6 messages that is actually relevant? :shrug:

Comment (pchaigno) [reply bpf/bpf_host.c:248]: Ok. Packet asymmetry is usually only an issue for service handling and I'm guessing that won't be an issue with ICMPv6 packets. So probably okay given no user impact?

Comment (julianwiedmann) [reply bpf/bpf_host.c:248]: > So probably okay given no user impact?

That was my gut feeling, yes. Will leave it as first patch in the split-up series then, so that we can easily cherry-pick it later if necessary.

**Pull request closed without merging**
